release-notes.txt: add 2024.07 release notes (#20797)
RIOT-OS/RIOT: release-notes.txt

access, independent of the underlying hardware (with partial POSIX compliance).

RIOT is developed by an international open-source community which is
independent of specific vendors (e.g. similarly to the Linux community) and is
licensed with a non-viral copyleft license (LGPLv2.1), which allows indirect
business models around the free open-source software platform provided by RIOT.

About this release
==================

The 2024.07 release includes:

## List of notable changes here ##

175 pull requests, composed of 358 commits, have been merged since the
last release, and 5 issues have been solved. 36 people contributed with
code in 80 days. 650 files have been touched with 88156 (+) insertions and
3949 deletions (-).

Notations used below
====================

    + means new feature/item
    * means modified feature/item
    - means removed feature/item

New features and changes
========================

Core
----

* core/assert: halt running thread instead of panic (#20627)
* core/lib/assert: consistify output and cleanup documentation (#20701)
* core/panic: make reboot on panic configurable (#20630)

> * core/thread: "fix" valgrind erros in thread_measure_stack_free() (#20726)

Reviewer: ```suggestion
* core/thread: "fix" valgrind errors in thread_measure_stack_free() (#20726)
```

I already updated the PR title to match this.

+ sys/shell: add support for running a batch of commands from a file (#20651)
* rust: bump rust-riot-{sys,wrappers} version (#20786)
* shell/vfs: Handle print failure in genfile cmd gracefully (#20595)
* sys/hashes: change SHA2 PAD type to const (#20729)
* sys/psa_crypto: ed25519 private key {ex,im}port (#20334)
* sys/psa_crypto: Implement persistent key storage (#20099)
* sys/psa_crypto: remove keysize determination from Kconfig files (#20674)
* sys/psa_crypto: SHA-{384,512/{224,256}} (#20598)
* sys/psa_crypto: sha3 support (#20698)
* sys/shell: Update cryptoauthlib shell commands (#20673)
* sys/suit: return error when URL buffer is too small [backport
  2024.01] (#20563)
* sys/vfs: use atomic_utils rather C11 atomics (#20513)
* treewide: Rust updates [backport 2024.01] (#20384)

Networking
----------

+ gcoap:  add coaps forward proxy (#20454)
+ gnrc, nanocoap: add optional work-arounds for buggy CoAP servers (#20564)
+ nanocoap_fs: add nanoCoAP as VFS backend (remote CoAP fs) (#20687)
+ nanocoap_sock: add nanocoap_get_blockwise_to_buf() (#20690)
+ nanocoap_sock: implement separate response (#20266)
* doc: fix links to GNRC master thesis (#20785)

> * driver/w5500: driver for the W5500 ethernet chip (#20301)

Reviewer: This driver is IMO noteworthy.

* Makefile.include: creation of CACHEDIR.TAG as a dependency of pkg-
  prepare (#20689)
* pkg/littlefs2: bump to v2.9.2 (#20591)
* pkg/tflite-micro: Update tflite-micro to latest version. requires
  update of pkg/flatbuffers as well. (#20683)
* pkg/wamr: updated package to WAMR 1.3.3 (#20691)

Boards
------

+ boards/arduino-nano-33-ble-sense: add support for arduino-
  nano-33-ble-sense (#20668)
+ boards/b-l072z-lrwan1: add missing adc feature (#20733)
+ boards/nucleo-f303ze: add pinout (#20717)
+ boards/nucleo-l452re: add ADC (#20778)
+ boards/sodaq-sara-sff: add status pin for SARA (#20279)
+ cpu/stm32: add ADC support for WB55 (#20773)
+ drivers/led: add LED_NUMOF and convenience inline functions (#20783)
+ drivers/ws281x: Add saul support (#20562)
* boards/common/stm32: simplify ifdef logic (#20617)
* boards/nRF52xx: correctly set JLINK_DEVICE (#20599)
* boards/nucleo-l476rg & stm32l476g-disco: fix MCU table (#20751)
* boards/stm32l0538-disco: enable cpy2remed (#20766)
* boards/stm32l476g-disco: enable cpy2remed (#20734)
* build system: clean up netif features (#20700)
* cpu/{gd32v,stm32}/periph/adc: make ADC clock setable (#19630)

CPU
---

> + cpu/cortexm: add stack limit support for Cortex-M33 (#20633)

Reviewer: This is also IMO noteworthy

* cpu/cortexm: rework bkpt instruction call on panic (#20616)
* cpu/cortexm_common: print last active thread on stack corruption (#20709)
* cpu/esp: Handle format print errors (#20596)

> * cpu/msp430: implement power management (#20613)

Reviewer: The MSP430 port has actually matured quite a bit over the last few releases. There is still some way to get parity with the level of support we have for Cortex M MCUs such as STM32 or nRF52.

Might be worth a short note, but the gist would be "a number of cleanups, fixes and implementation of missing have accumulated over the past releases"; this PR is not more relevant than the other MSP430 PRs.

* cpu/nrf5x: implement pm_off() for nRF53/9160 (#20590)
* cpu/nrf5x_common: fix ztimer issue on warm-boot (#20665)
* cpu/nrf5x_common: properly calibrate RC-based low-frequency clock (#20669)
* cpu/sam0: Remove deprecated muxpos (#20525)
* cpu/sam0_common: avoid bitfield usage (#20747)
* cpu/samd21: avoid the use of bitfield (#20713)
* cpu/samd5x: avoid the use of bitfields (#20714)
* cpu/samd5x: handle CAN errors (#20667)
* cpu/saml1x: avoid the use of bitfield in register calls (#20710)

> * cpu/saml21: avoid the use of bitfield in register call (#20708)

Reviewer: This is IMO worth a note: The first steps towards updating the vendor header files and allowing the port of new SAM* hardware has been made by @dylad.

* cpu/stm32/periph/stm32_eth: provide confirm_send (#18428)
* cpu/stm32: de-duplicate ifdefs (#20609)
* periph/flashpage: remove deprecated flashpage_*_free functions (#18093)
* pkg/lwip: make use of confirm send (#18427)
* Remove extraneous bitwise assignment operation on NRF5X_common gpio.c (#20737)
* sam0_eth: implement .confirm_send() to fix fragmented sending (#20666)
* tree wide: compilation fixes for `native64` on musl systems (#20730)

Device Drivers
--------------

+ drivers/at24cxxx: Add M24C01 device and enhance documentation (#20588)
+ drivers/led: add LEDX_IS_PRESENT defines (#20637)

> + drivers/mtd: Add check for interger overflow (#20587)

Reviewer: ```suggestion
+ drivers/mtd: Add check for integer overflow (#20587)
```

(I already fixed the typo in the PR to match the release notes.)

+ Drivers: add MAX31855 thermocouple-to-digital converter (#20741)

> * drivers/at86rf215: port to new netdev API (#20672)

Reviewer: This and the `stm32_eth` driver have been the first that have been converted. This may also be worth a note, as those are the first drivers that have been ported to the new API.

The blocker so far was lwIP, which has been stalling the conversion to the new API for ... two years or so?.

Support for the new netdev API has been added to lwIP in this release which allowed the first drivers to be converted.

There are two cool things this allows: The first is that the new API is easier to implement by drivers as the API more closely matches the behavior of real world hardware. The second is that the network stack now actually can synchronize with the netdev, which would allow maximizing TX utilization.

Author: Is there some place to read more about the differences between the old and the new netdev API? I've looked at the docs but couldn't find any.

Reviewer: https://doc.riot-os.org/structnetdev__driver.html#a39585137953a92c1c58b5e0ad3262096 and https://doc.riot-os.org/structnetdev__driver.html#aacad0fea752b3a0ec9beb30470c4f34e explain both variants.

The plan back then was that netdevs are quickly converted with little fuzz, and backward compatibility with the old variant is dropped quickly after. But the PR to update lwIP to support both variants got stalled for years...

Anyway, with that resolved, it would indeed be nice to just quickly convert all netdevs to the new API.

Author: So maybe a good goal for the next release, and report success in the next release notes? :)

Reviewer: Indeed. I hope I can find some time for that :)

* drivers/at: expose some internal API (#20702)

> * drivers/periph_wdt: fix auto starting of watchdog timer (#20755)

Reviewer: I'd move this to bug fixes.

Documentation
-------------

+ boards/nucleo-c031c6: add pinout to doc (#20657)
+ boards/nucleo-l452re: add pinout to doc page (#20678)
+ boards/nucleo-l4r5zi and l496zg: add pinout to documentation page (#20707)
+ boards/nucleo144: add pinout diagrams to boards documentation (#20614)
+ doc/getting-started: add hint for correct access rights on serial
  device (#20600)
+ doc/terminals: add putty and sort terminals (#20611)
* /README.md: match doc (#20740)
* boards/doc: Adafruit Feather nRF52840 Sense is popular (#20597)
* boards/nucleo-f429zi & f439zi: documentation improvements (#20594)
* boards/nucleo-f722ze: doc update (#20671)
* boards/stm32l0538-disco: doc improvements (#20767)
* boards: Update SAM0-based documentation (#20621)
* doc/doxygen: remove default value for search form (#20681)

Build System / Tooling
----------------------

+ pkg: Add script to check third party package version (#20662)
+ tools/dhcpv6-pd_ia: fix: Add explicit subnet id (#20776)
* dist/tools/buildsystem_sanity_check: make shellcheck happy (#20721)
* dist/tools/tapsetup: make compatible with doas (#20712)
* Makefile.include: place compile_commands.json by rule target (#20716)

> * makefiles/docker.inc.mk: Pin riotbuild version with BUILD_IN_DOCKER=1 (#20472)

Author: Maybe worth a highlighting note?

Reviewer: Yes, that seems to be something users of docker are better aware of.

* makesfiles/jlink: fix exports for flashing (#20779)

Examples
--------

* examples/psa_crypto: pass environment variable down to docker (#20663)

Testing
-------

+ tests/drivers/at: add check if device is initialized before sending
  command (#20140)
* check-labels.yml: bump to v1.1.2 (#20727)
* ci: disable esp32-wroom-32 tests (#20715)
* release-tests.yaml: remove (always failing) matrix notification (#20742)
* test/sys/progress_bar: temporary printout to debug flaky test (#20763)
* tests/drivers/pn532: fix init error handling and increase verbosity (#20750)
* tests/sys/progress_bar: move configuration to app.config (#20764)
* tests/sys/progress_bar: remove debug print (#20768)
* tests/sys/psa_crypto*: remove test symlinks (#20545)
* tests/sys/psa_crypto: increase timeout (#20612)
* tests/sys/psa_crypto: test interleaved hash operations (#20606)
* tests/sys/psa_crypto_ecdsa: fix stacksize determination (#20760)

API Changes
-----------

* build system: rename ethernet feature into netif_ethernet (#20694)
* cpu/native: fix build with musl (#18942)
* pkg/monocypher: bump to 4.0.2 (#20670)

Uncategorized
-------------

> + .editorconfig: add EditorConfig (#20746)

Author: Move to tooling

> + release-notes.txt: add 2024.04 release notes (#20641)

Author: ```suggestion
```

And 32 minor changes.

> Deprecations
> ============
> 
> Deprecations (1)
> ----------------

Author: ```suggestion
Deprecations (1)
================
```

* boards/waspmote-pro: deprecate board (#20649)

Bug fixes (22)
==============

* build system: add netif_openwsn feature (#20695)
* makefiles: add OPENOCD_DEBUG_ADAPTER and SLOT_AUX_LEN to riotboot
  build environment (#20752)

> * tree-wide: Introduce netif feature and use it (#20682)

Author: Move to tooling

business models around the free open-source software platform provided by RIOT.

About this release
==================

The 2024.07 release, nicknamed "StrikinglyCrowded", includes many features and bug fixes.
Here are some of the highlights:

+ RIOT now boots on the Arduino Nano 33 BLE Sense (#20668)
+ To support boards with Ethernet plugs such as the Adafruit Ethernet FeatherWing, RIOT now ships a driver for the W5500 ethernet chip (#20301)
+ RIOT now leverages the ARMv8-M hardware support to enforce the stack size (#20633)
+ The PSA Crypto API implementation in RIOT now supports persistent key storage and additional hashing algorithms (#20099, #20598, #20698)
+ This release features several useful additions to the RIOT CoAP stacks: a coaps forward proxy for gcoap, delayed responses for nanocoap_sock, and the ability to mount a remote filesystem over CoAP (#20454, #20266, #20687)
* Support of the 16-bit MSP430 MCU has been improving over the last few releases, with the addition of power management in this release as an example (#20613)
* First steps towards enabling RIOT support of new Microchip SAM MCUs (#20708, #20710, #20713, #20714)

> * Users of BUILD_IN_DOCKER=1 will now be warned if they are using an outdated docker container (#20472)

Reviewer: ```suggestion
* Compiling with `BUILD_IN_DOCKER=1` will now by default use the exact container version the code was tested in the CI, fetching it from upstream if not locally present. (#20472)
```

I forgot to update the PR description. The commit description does indeed match the behavior.

3949 deletions (-).

Notations used below
====================

    + means new feature/item
    * means modified feature/item
    - means removed feature/item

New features and changes
========================

Core
----

* core/assert: halt running thread instead of panic (#20627)
* core/lib/assert: consistify output and cleanup documentation (#20701)
* core/panic: make reboot on panic configurable (#20630)
* core/thread: "fix" valgrind errors in thread_measure_stack_free() (#20726)

System Libraries
----------------

+ sys/event.h: Fixed documentation error by adding a type cast. (#20704)
+ sys/shell: add support for running a batch of commands from a file (#20651)
* rust: bump rust-riot-{sys,wrappers} version (#20786)
* shell/vfs: Handle print failure in genfile cmd gracefully (#20595)
* sys/hashes: change SHA2 PAD type to const (#20729)
* sys/psa_crypto: ed25519 private key {ex,im}port (#20334)
* sys/psa_crypto: Implement persistent key storage (#20099)
* sys/psa_crypto: remove keysize determination from Kconfig files (#20674)
* sys/psa_crypto: SHA-{384,512/{224,256}} (#20598)
* sys/psa_crypto: sha3 support (#20698)
* sys/shell: Update cryptoauthlib shell commands (#20673)

> * sys/suit: return error when URL buffer is too small [backport

Reviewer 2: Maybe it should just be 20559 not the backport?

Author: Good catch, that was actually part of the last release already.

  2024.01] (#20563)
* sys/vfs: use atomic_utils rather C11 atomics (#20513)

> * treewide: Rust updates [backport 2024.01] (#20384)

Reviewer 2: Same thing here: 20303

Author: And same here, too. I wonder why they ended up in the automatically generated changelog.

business models around the free open-source software platform provided by RIOT.

About this release
==================

The 2024.07 release includes:

+ RIOT now boots on the Arduino Nano 33 BLE Sense (#20668)
+ To support boards with Ethernet plugs such as the Adafruit Ethernet FeatherWing, RIOT now ships a driver for the W5500 ethernet chip (#20301)
+ RIOT now leverages the ARMv8-M hardware support to enforce the stack size (#20633)
* The PSA Crypto API implementation in RIOT now supports persistent key storage and additional hashing algorithm (#20099, #20598, #20698)

> * Support of the 16-bit MSP430 MCU has been improving over the last few releases, with the addition of power managment in this release as an example (#20613)

Reviewer 3: ```suggestion
* Support of the 16-bit MSP430 MCU has been improving over the last few releases, with the addition of power management in this release as an example (#20613)
```

> * First steps towards enabling RIOT support of new Atmel SAM MCUs (#20708, #20710, #20713, #20714)

Reviewer 4: ```suggestion
* First steps towards enabling RIOT support of new Microchip SAM MCUs (#20708, #20710, #20713, #20714)
```

Author: Whoopsie :sweat_smile:

----

* core/assert: halt running thread instead of panic (#20627)
* core/lib/assert: consistify output and cleanup documentation (#20701)
* core/panic: make reboot on panic configurable (#20630)
* core/thread: "fix" valgrind errors in thread_measure_stack_free() (#20726)

System Libraries
----------------

+ sys/event.h: Fixed documentation error by adding a type cast. (#20704)
+ sys/shell: add support for running a batch of commands from a file (#20651)
* rust: bump rust-riot-{sys,wrappers} version (#20786)
* shell/vfs: Handle print failure in genfile cmd gracefully (#20595)
* sys/hashes: change SHA2 PAD type to const (#20729)
* sys/psa_crypto: ed25519 private key {ex,im}port (#20334)
* sys/psa_crypto: Implement persistent key storage (#20099)
* sys/psa_crypto: remove keysize determination from Kconfig files (#20674)
* sys/psa_crypto: SHA-{384,512/{224,256}} (#20598)
* sys/psa_crypto: sha3 support (#20698)
* sys/shell: Update cryptoauthlib shell commands (#20673)
* sys/suit: return error when URL buffer is too small [backport
  2024.01] (#20563)
* sys/vfs: use atomic_utils rather C11 atomics (#20513)
* treewide: Rust updates [backport 2024.01] (#20384)

Networking
----------

> + gcoap:  add coaps forward proxy (#20454)

Reviewer 3: This is a noteworthy feature

+ gnrc, nanocoap: add optional work-arounds for buggy CoAP servers (#20564)

> + nanocoap_fs: add nanoCoAP as VFS backend (remote CoAP fs) (#20687)

Reviewer 3: Not sure how useful this is in general, but it allows to mount a remote fs that is exported via `nanocoap_fileserver` (or `aiocoap-fileserver`), so might be an interesting feature.

+ nanocoap_sock: add nanocoap_get_blockwise_to_buf() (#20690)

> + nanocoap_sock: implement separate response (#20266)

Reviewer 3: separate response is also a useful new feature for the nanoCoAP server

+ sys/event.h: Fixed documentation error by adding a type cast. (#20704)
+ sys/shell: add support for running a batch of commands from a file (#20651)
* rust: bump rust-riot-{sys,wrappers} version (#20786)
* shell/vfs: Handle print failure in genfile cmd gracefully (#20595)
* sys/hashes: change SHA2 PAD type to const (#20729)
* sys/psa_crypto: ed25519 private key {ex,im}port (#20334)
* sys/psa_crypto: Implement persistent key storage (#20099)
* sys/psa_crypto: remove keysize determination from Kconfig files (#20674)
* sys/psa_crypto: SHA-{384,512/{224,256}} (#20598)
* sys/psa_crypto: sha3 support (#20698)
* sys/shell: Update cryptoauthlib shell commands (#20673)
* sys/suit: return error when URL buffer is too small [backport
  2024.01] (#20563)
* sys/vfs: use atomic_utils rather C11 atomics (#20513)
* treewide: Rust updates [backport 2024.01] (#20384)

Networking
----------

+ gcoap:  add coaps forward proxy (#20454)
+ gnrc, nanocoap: add optional work-arounds for buggy CoAP servers (#20564)
+ nanocoap_fs: add nanoCoAP as VFS backend (remote CoAP fs) (#20687)
+ nanocoap_sock: add nanocoap_get_blockwise_to_buf() (#20690)
+ nanocoap_sock: implement separate response (#20266)

> * doc: fix links to GNRC master thesis (#20785)

Reviewer 5: A bug fix at most, but not sure this is noteworthy change for the change log ;-).
